Action/PostItemAction.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+/*
+ * This file is part of the DunglasApiBundle package.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Dunglas\ApiBundle\Action;
+
+use Dunglas\ApiBundle\Event\DataEvent;
+use Dunglas\ApiBundle\Event\Events;
+use Dunglas\ApiBundle\Exception\RuntimeException;
+use Dunglas\ApiBundle\Exception\ValidationException;
+use Dunglas\ApiBundle\Model\DataProviderInterface;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\Serializer\SerializerInterface;
+use Symfony\Component\Validator\Validator\ValidatorInterface;
+
+/**
+ * Add a new resource to a collection.
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+class PostItemAction
+{
+    use ActionUtilTrait;
+
+    /**
+     * @var DataProviderInterface
+     */
+    private $dataProvider;
+    /**
+     * @var SerializerInterface
+     */
+    private $serializer;
+    /**
+     * @var ValidatorInterface
+     */
+    private $validator;
+    /**
+     * @var EventDispatcherInterface
+     */
+    private $eventDispatcher;
+
+    public function __construct(DataProviderInterface $dataProvider, SerializerInterface $serializer, ValidatorInterface $validator, EventDispatcherInterface $eventDispatcher)
+    {
+        $this->dataProvider = $dataProvider;
+        $this->serializer = $serializer;
+        $this->eventDispatcher = $eventDispatcher;
+        $this->validator = $validator;
+    }
+
+    /**
+     * Create a new item.
+     *
+     * @param Request $request
+     *
+     * @return mixed
+     *
+     * @throws RuntimeException
+     */
+    public function __invoke(Request $request)
+    {
+        list($resourceType, $format) = $this->extractAttributes($request);
+
+        $data = $this->serializer->deserialize(
+            $request->getContent(),
+            $resourceType->getEntityClass(),
+            $format,
+            $resourceType->getDenormalizationContext()
+        );
+
+        $this->eventDispatcher->dispatch(Events::PRE_CREATE_VALIDATION, new DataEvent($resourceType, $data));
+
+        $violations = $this->validator->validate($data, null, $resourceType->getValidationGroups());
```

Comment:
Here it's very similar IMO: despite the name (`$validator->validate()`), the validation is only triggered in the action layer but is done in the domain layer. The Symfony **validator** will successively call all your customs (or builtin) **validators** (in config files or with annotations or entities) and returns a list of validation failure (the payload of the exception here). Then exception is caught by a responder that will convert it in Hydra error.

IMO we can say that validation is done in the domain layer here. It is only triggered in the action but all the validation logic reside in the domain (and if no validator are defined, it will do nothing). 

An alternative solution can be to move the triggering itself in the domain layer (in `DataProvider`) but it has a big drawback: validation will be triggered in all cases (even in `GET ` methods). That mean a performance decrease and possible unwanted side effects in case data retrieved from the persistence system violate some validation rules (this is really bad practice but it happens).

I'm not sure of what to do here. Throwing an event have other another advantage: it allows to use easily an alternative validator that the Symfony one.